doc/designs/quic-design/dgram-api.md

```diff
@@ -0,0 +1,487 @@
+Datagram BIO API revisions for sendmmsg/recvmmsg
+================================================
+
+We need to evolve the API surface of BIO which is relevant to BIO_dgram (and the
+eventual BIO_dgram_mem) to support APIs which allow multiple datagrams to be
+sent or received simultaneously, such as sendmmsg(2)/recvmmsg(2).
+
+The adopted design
+------------------
+
+### Design decisions
+
+The adopted design makes the following design decisions:
+
+- We use a sendmmsg/recvmmsg-like API. The alternative API was not considered
+  for adoption because it is an explicit goal that the adopted API be suitable
+  for concurrent use on the same BIO.
+
+- We define our own structures rather than using the OS's `struct mmsghdr`.
+  The motivations for this are:
+
+  - It ensures portability between OSes and allows the API to be used
+    on OSes which do not support `sendmmsg` or `sendmsg`.
+
+  - It allows us to use structures in keeping with OpenSSL's existing
+    abstraction layers (e.g. `BIO_ADDR` rather than `struct sockaddr`).
+
+  - We do not have to expose functionality which we cannot guarantee
+    we can support on all platforms (for example, arbitrary control messages).
+
+  - It avoids the need to include OS headers in our own public headers,
+    which would pollute the environment of applications which include
+    our headers, potentially undesirably.
+
+- For OSes which do not support `sendmmsg`, we emulate it using repeated
+  calls to `sendmsg`. For OSes which do not support `sendmsg`, we emulate it
+  using `sendto` to the extent feasible. This avoids the need for code consuming
+  these new APIs to define a fallback code path.
+
+- We do not define any flags at this time, as the flags previously considered
+  for adoption cannot be supported on all platforms (Win32 does not have
+  `MSG_DONTWAIT`).
+
+- We ensure the extensibility of our `BIO_MSG` structure in a way that preserves
+  ABI compatibility using a `stride` argument which callers must set to
```

Comment:
I'm not sure I like this approach.

What about a call to allocate an array of BIO_MSG structures and calls to populate/retrieve their field fields?
We'd be able to keep the BIO_MSG internals opaque and avoid a previous nasty problem.  API and ABI preserved.  Having these allocated but reusable seems like a good compromise (via setter functions).

An API preserving alternative would be to make the structure public and have any change to it be an ABI break -- i.e. major releases only.

Reply:
IMO we are allocating too much already. We should avoid that for this kind of low level interface.

Reply:
I think that would be really overkill and we don't want to force heap allocation of these structures, since they will only be needed transiently.

We will need to add stuff to this structure later e.g. for ECN, maybe ToS byte, etc. I don't think "we can never change this" is an option.

A stride argument works well and isn't an uncommon solution. They're commonly used to index rows in image buffers, and the Win32 API is full of "pass the structure size" as an extensibility mechanism. It may be relatively novel for OpenSSL but it seems like a simple and effective solution to me.